### PR TITLE
Remove @with, add @attach

### LINKED ISSUE #364: `@with` macro clashes with `Base.@with` in Julia 1.11+
Julia 1.11 is the first version to export `Base.@with`:
```
help?> @with
  @with vars... expr

  Macro version of with(f, vars...) but with expr instead of f function. This is similar to using with with a do block, but avoids creating a closure.
```

This causes a name conflict with `DataFramesMeta.@with` (i.e., error when the latter used in a module). MWE:
```julia
julia> module Foo
    using DataFrames, DataFramesMeta

    const df = DataFrame(a = [1, 2, 3])
    print() = @with df println.(:a)
end
WARNING: both DataFramesMeta and Base export "@with"; uses of it in module Foo must be qualified
ERROR: LoadError: UndefVarError: `@with` not defined
in expression starting at REPL[14]:6

```

### PR DESCRIPTION
**Pull request by pdeffebach**
fixes #364

### CODE REVIEW

File: src/macros.jl [hunk 1]
```diff
@@ -663,6 +663,17 @@ macro attach(d, body)
     esc(attach_helper(d, body))
 end
 
+"""
+    @with(x, args...)
+
+Deprecated version of `@attach`, see `?@with` for details.
+"""
+macro with(x, args...)
+    @warn "`@with is deprecated, use `@with` instead."
```
Comment (bkamins) [inline src/macros.jl:672]:
```suggestion
    @warn "`@with is deprecated, use `@attach` instead."
```

File: src/macros.jl [hunk 2]
```diff
@@ -663,6 +663,17 @@ macro attach(d, body)
     esc(attach_helper(d, body))
 end
 
+"""
+    @with(x, args...)
+
+Deprecated version of `@attach`, see `?@with` for details.
```
Comment (bkamins) [inline src/macros.jl:669]:
```suggestion
Deprecated version of `@attach`, see `?@attach` for details.
```

**Comment by pdeffebach** [comment]
Tests pass

**Comment by bkamins** [comment]
There are some doctest failures

**Review by bkamins — APPROVED** [review]

**Comment by pdeffebach** [comment]
This can be a small version bump, right? Since it just deprecates, we don't need to make a `0.15.0` version.

**Comment by bkamins** [comment]
we add new `@attach`. In the past, we would e.g. in DataFrames.jl make 0.15 release. But recently people were saying that such minor changes are OK to be made in 0.14.2.

**Comment by pdeffebach** [comment]
Okay I'm just going to merge this with a minor version bump. I don't want to release too many breaking changes and I want people to see the deprecation. 

I will also announce on discourse soon.

**Comment by bkamins** [comment]
OK

**Comment by nalimilan** [comment]
Sorry, for some reason I missed this. It's really unfortunate that Julia exports this general name now. Have you considered calling it `@within` instead of `@attach` (which indeed has a bad reputation in R as it attaches copies of columns to the global environment)?

**Comment by pdeffebach** [comment]
`@within` is a good name. I guess I still prefer `@attach` since

1. It's a verb, kind of like `@select`, `@transform` etc. 
2. I like the idea of us re-doing an R feature, but "correctly", without the scoping and performance drawbacks of `attach`.

**Comment by pdeffebach** [comment]
Closing this due to https://github.com/JuliaLang/julia/pull/53004#issuecomment-1981109599 being merged

**Pull request closed without merging**